src/helm/proxy/clients/palmyra_client.py

```diff
@@ -0,0 +1,125 @@
+            data=json.dumps(raw_request),
+        )
+        result = json.loads(response.text)
+        if "error" in result:
+            raise ValueError(f"Request failed with error: {result['error']}")
+        return result
+
+    def make_request(self, request: Request) -> RequestResult:
+        """Make a request"""
+        raw_request = {
+            "prompt": request.prompt,
+            "maxTokens": request.max_tokens,
+            "temperature": request.temperature,
+            "topP": request.top_p,
+            "bestOf": request.top_k_per_token,
+            "stop": request.stop_sequences,
+            "random_seed": request.random,
+        }
+
+        completions: List[Sequence] = []
+        model_name: str = request.model_engine
+
+        # `num_completions` is not supported, so instead make `num_completions` separate requests.
+        for completion_index in range(request.num_completions):
+            try:
+
+                def do_it():
+                    # HACKY: Use the random seed to get different results for each completion.
+                    raw_request["random_seed"] = (
+                        completion_index if request.random is None else request.random + completion_index
+                    )
+                    result = self._send_request(model_name, raw_request)
+                    if "choices" not in result:
+                        raise ValueError(f"Invalid response: {result}")
+                    return result
+
+                # We need to include the engine's name to differentiate among requests made for different model
+                # engines since the engine name is not included in the request itself.
+                # In addition, we want to make `request.num_completions` fresh
+                # requests, cache key should contain the completion_index.
+                # Echoing the original prompt is not officially supported by Writer. We instead prepend the
+                # completion with the prompt when `echo_prompt` is true, so keep track of it in the cache key.
+                cache_key = Client.make_cache_key(
+                    {
+                        "engine": request.model_engine,
+                        "echo_prompt": request.echo_prompt,
+                        "completion_index": completion_index,
+                        **raw_request,
+                    },
+                    request,
+                )
+
+                response, cached = self.cache.get(cache_key, wrap_request_time(do_it))
+            except (requests.exceptions.RequestException, AssertionError) as e:
+                error: str = f"PalmyraClient error: {e}"
+                return RequestResult(success=False, cached=False, error=error, completions=[], embedding=[])
+
+            # Post process the completion.
+            response_text: str = response["choices"][0]["text"]
+            # NOTE(josselin): The Writer API sometimes returns a space at the beginning of the completion.
```

Comment:
Arghhh.

Could open a GitHub issue to document all of the post-processing string hacks, and then add a `# TODO(#github issue number): ...` comment?

Reply:
Done: https://github.com/stanford-crfm/helm/issues/1512